ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,186 @@
+        lastSampleTime = time.milliseconds();
+        previousPollTime = new HashMap<>();
+        previousRestoreConsumerPollTime = new HashMap<>();
+        previousSendTime = new HashMap<>();
+        previousFlushTime = new HashMap<>();
+    }
+
+    // for testing
+    public UtilizationMetricsListener(final long windowSize, final List<KafkaStreams> streams) {
+        this.kafkaStreams = streams;
+        this.metrics = new LinkedList<>();
+        // we can add these here or pass it in through the constructor
+        metrics.add("poll-time-total");
+        metrics.add("restore-consumer-poll-time-total");
+        metrics.add("send-time-total");
+        metrics.add("flush-time-total");
+        time = Time.SYSTEM;
+        this.windowSize = windowSize;
+        lastSampleTime = time.milliseconds() - 100L;
+        previousPollTime = new HashMap<>();
+        previousRestoreConsumerPollTime = new HashMap<>();
+        previousSendTime = new HashMap<>();
+        previousFlushTime = new HashMap<>();
+    }
+
+    @Override
+    public void onCreate(
+            final ServiceContext serviceContext,
+            final MetaStore metaStore,
+            final QueryMetadata queryMetadata) {
+        kafkaStreams.add(queryMetadata.getKafkaStreams());
+    }
+
+    @Override
+    public void onDeregister(final QueryMetadata query) {
+        kafkaStreams.remove(query.getKafkaStreams());
+        previousPollTime.remove("poll-time-total");
+        previousRestoreConsumerPollTime.remove("restore-consumer-poll-time-total");
+        previousSendTime.remove("send-time-total");
+        previousFlushTime.remove("flush-time-total");
+    }
+
+    @Override
+    public void run() {
+        logger.info("Reporting CSU thread level metrics");
+        while (true) {
+            logger.info("the current processing ratio is " + processingRatio() + "%");
+            try {
+                Thread.sleep(windowSize);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    private void reportSystemMetrics() {
+        try {
+            BufferedReader br = new BufferedReader(new FileReader("/procfs/sys/1/stat"));
+            String[] columns = br.readLine().split(" ");
+            String timeUser = columns[13];
+            String timeKernel = columns[14];
+            int clockTicks = 100;
+
+            logger.info("we've got " + timeUser + " for timeuser, " + timeKernel + " for timekernel");
+        } catch(Exception e) {
+            logger.info("something went wrong finding cpu utilization metrics " + e.getMessage());
+        }
+
+        logger.info("we're using some disk");
+    }
+
+    // public for testing
+    public double processingRatio() {
+        long sampleTime = time.milliseconds();
+        double blockedTime = sampleTime - lastSampleTime;
+
+        final long windowSize = Math.max((sampleTime - lastSampleTime), this.windowSize);
```

Comment:
why are you taking a max with the windowSize here? If it's to catch the case where sampleTime < lastSampleTime it would be more accurate to just return 0 or throw - something has definitely gone wrong and we shouldn't just compute a value that uses the requested windowSize.

Reply:
That was something I mostly did for testing - to make sure we either getting a window size that's bigger than the set window size because the thread is slow or we're getting the window size, but nothing smaller. I removed this when I did the test clean up tho